x-pack/plugins/inference/server/chat_complete/adapters/gemini/gemini_adapter.ts

```diff
@@ -0,0 +1,234 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import * as Gemini from '@google/generative-ai';
+import { from, map, switchMap } from 'rxjs';
+import { Readable } from 'stream';
+import type { InferenceConnectorAdapter } from '../../types';
+import { Message, MessageRole } from '../../../../common/chat_complete';
+import { ToolChoiceType, ToolOptions } from '../../../../common/chat_complete/tools';
+import type { ToolSchema, ToolSchemaType } from '../../../../common/chat_complete/tool_schema';
+import { eventSourceStreamIntoObservable } from '../../../util/event_source_stream_into_observable';
+import { processVertexStream } from './process_vertex_stream';
+import type { GenerateContentResponseChunk } from './types';
+
+/**
+ * We need to use the connector's format, not directly Gemini's...
+ *
+ * In practice, either use `content` that will be automatically converted to parts,
+ * or use `parts` directly for advanced usages such as tool response.
+ *
+ * See x-pack/plugins/stack_connectors/server/connector_types/gemini/gemini.ts
+ */
+interface GeminiMessage {
+  role: 'assistant' | 'user';
+  content?: string;
+  parts?: Gemini.Part[];
+}
+
+interface GeminiToolConfig {
+  mode: 'AUTO' | 'ANY' | 'NONE';
+  allowedFunctionNames?: string[];
+}
+
+export const geminiAdapter: InferenceConnectorAdapter = {
+  chatComplete: ({ executor, system, messages, toolChoice, tools }) => {
+    return from(
+      executor.invoke({
+        subAction: 'invokeStream',
+        subActionParams: {
+          messages: messagesToGemini({ system, messages }),
+          tools: toolsToGemini(tools),
+          toolConfig: toolChoiceToConfig(toolChoice),
```

Comment:
Function calling is support on [gemini-1.0-pro, gemini-1.5-flash-latest and gemini-1.5-pro-latest](https://ai.google.dev/gemini-api/docs/function-calling#single-turn-using-mode-and-allowed-functions), so basically all version but the `-vision` one which is AFAIK very specific, and discontinued in ~ 6 months anyway, so I felt like it was better to implement function calling the right way now rather than re-using the fake function invocation hack that o11y was using.